File: frontend/src/game/leaderboardTrack.js
/**
 * Rolling 1000-step window for server leaderboard (independent of stats reset).
 * See docs/design/13-player-statistics.md section 7.7.
 */

import { explorationSteps } from './playerStatistics.js'

export const LEADERBOARD_WINDOW_STEPS = 1000
export const LEADERBOARD_MIN_LIFETIME_STEPS = 1000
export const LEADERBOARD_DISPLAY_SCALE = 100

/**
 * @typedef {{ steps: number, gold: number, xp: number }} LeaderboardTrackSegment
 * @typedef {{ lifetimeSteps: number, segments: LeaderboardTrackSegment[] }} LeaderboardTrack
 */

/** @returns {LeaderboardTrack} */
export function createEmptyLeaderboardTrack() {
  return { lifetimeSteps: 0, segments: [] }
}

/** @param {unknown} raw @returns {LeaderboardTrack} */
export function normalizeLeaderboardTrack(raw) {
  const empty = createEmptyLeaderboardTrack()
  if (!raw || typeof raw !== 'object') return empty
  const o = /** @type {Record<string, unknown>} */ (raw)
  /** @type {LeaderboardTrackSegment[]} */
  const segments = []
  if (Array.isArray(o.segments)) {
    for (const seg of o.segments) {
      if (!seg || typeof seg !== 'object') continue
      const s = /** @type {Record<string, unknown>} */ (seg)
      const steps = Math.max(0, Math.floor(Number(s.steps) || 0))
      if (steps <= 0) continue
      segments.push({
        steps,
        gold: Math.max(0, Number(s.gold) || 0),
        xp: Math.max(0, Number(s.xp) || 0),
      })
    }
  }
  let lifetimeSteps = Math.max(0, Math.floor(Number(o.lifetimeSteps) || 0))
  if (lifetimeSteps <= 0 && segments.length > 0) {
    lifetimeSteps = segments.reduce((sum, seg) => sum + seg.steps, 0)
  }
  return trimWindow({ lifetimeSteps, segments })
}

/**
 * @param {LeaderboardTrack} track
 * @param {{ steps?: number, gold?: number, xp?: number }} delta
 * @returns {LeaderboardTrack}
 */
export function appendLeaderboardTrackSegment(track, delta) {
  const base = normalizeLeaderboardTrack(track)
  const steps = Math.max(0, Math.floor(Number(delta?.steps) || 0))
  if (steps <= 0) return base
  const segments = [
    ...base.segments,
    {
      steps,
      gold: Math.max(0, Number(delta?.gold) || 0),
      xp: Math.max(0, Number(delta?.xp) || 0),
    },
  ]
  return trimWindow({
    lifetimeSteps: base.lifetimeSteps + steps,
    segments,
  })
}

/**
 * @param {LeaderboardTrack} track
 * @param {{ combatActionSteps?: number, goldGained?: number, xpGained?: number }} battle
 */
export function applyBattleToLeaderboardTrack(track, battle) {
  return appendLeaderboardTrackSegment(track, {
    steps: battle?.combatActionSteps,
    gold: battle?.goldGained,
    xp: battle?.xpGained,
  })
}

/** @param {LeaderboardTrack} track @param {number} restStepsAdded */
export function applyRestToLeaderboardTrack(track, restStepsAdded) {
  return appendLeaderboardTrackSegment(track, {
    steps: restStepsAdded,
    gold: 0,
    xp: 0,
  })
}

/**
 * Bootstrap track from legacy saves that only had playerStats totals.
 * @param {LeaderboardTrack} track
 * @param {object} playerStats
 */
export function migrateLeaderboardTrackFromPlayerStats(track, playerStats) {
  const normalized = normalizeLeaderboardTrack(track)
  if (normalized.lifetimeSteps > 0) return normalized
  const steps = explorationSteps(playerStats)
  if (steps <= 0) return normalized
  return appendLeaderboardTrackSegment(createEmptyLeaderboardTrack(), {
    steps,
    gold: playerStats?.cumulativeGold,
    xp: playerStats?.cumulativeXp,
  })
}

/** @param {LeaderboardTrack} track */
export function windowTotals(track) {
  const normalized = normalizeLeaderboardTrack(track)
  let steps = 0
  let gold = 0
  let xp = 0
  for (const seg of normalized.segments) {
    steps += seg.steps
    gold += seg.gold
    xp += seg.xp
  }
  return { steps, gold, xp }
}

/** @param {LeaderboardTrack} track */
export function isLeaderboardEligible(track) {
  const normalized = normalizeLeaderboardTrack(track)
  return normalized.lifetimeSteps >= LEADERBOARD_MIN_LIFETIME_STEPS
}

/** @param {LeaderboardTrack} track */
export function leaderboardGoldPerStep(track) {
  const totals = windowTotals(track)
  if (totals.steps <= 0) return 0
  return totals.gold / LEADERBOARD_WINDOW_STEPS
}

/** @param {LeaderboardTrack} track */
export function leaderboardXpPerStep(track) {
  const totals = windowTotals(track)
  if (totals.steps <= 0) return 0
  return totals.xp / LEADERBOARD_WINDOW_STEPS
}

/**
 * @param {LeaderboardTrack} track
 * @returns {LeaderboardTrack}
 */
function trimWindow(track) {
  /** @type {LeaderboardTrackSegment[]} */
  const segments = track.segments.map((seg) => ({ ...seg }))
  let windowSteps = segments.reduce((sum, seg) => sum + seg.steps, 0)
  while (windowSteps > LEADERBOARD_WINDOW_STEPS && segments.length > 0) {
    const first = segments[0]
    const excess = windowSteps - LEADERBOARD_WINDOW_STEPS
    if (first.steps <= excess) {
      windowSteps -= first.steps
      segments.shift()
      continue
    }
    const keep = first.steps - excess
    const ratio = keep / first.steps
    segments[0] = {
      steps: keep,
      gold: first.gold * ratio,
      xp: first.xp * ratio,
    }
    windowSteps = LEADERBOARD_WINDOW_STEPS
  }
  return {
    lifetimeSteps: Math.max(0, Math.floor(Number(track.lifetimeSteps) || 0)),
    segments,
  }
}

File: internal/service/leaderboard_track.go
package service

import "encoding/json"

const (
	LeaderboardWindowSteps      = 1000
	LeaderboardMinLifetimeSteps = 1000
)

type leaderboardTrackSegment struct {
	Steps int     `json:"steps"`
	Gold  float64 `json:"gold"`
	Xp    float64 `json:"xp"`
}

type leaderboardTrack struct {
	LifetimeSteps int                       `json:"lifetimeSteps"`
	Segments      []leaderboardTrackSegment `json:"segments"`
}

func emptyLeaderboardTrack() leaderboardTrack {
	return leaderboardTrack{LifetimeSteps: 0, Segments: []leaderboardTrackSegment{}}
}

func explorationSteps(combat, rest int) int {
	total := combat + rest
	if total < 0 {
		return 0
	}
	return total
}

func normalizeLeaderboardTrack(raw json.RawMessage) leaderboardTrack {
	if len(raw) == 0 {
		return emptyLeaderboardTrack()
	}
	var track leaderboardTrack
	if err := json.Unmarshal(raw, &track); err != nil {
		return emptyLeaderboardTrack()
	}
	if track.LifetimeSteps < 0 {
		track.LifetimeSteps = 0
	}
	if track.Segments == nil {
		track.Segments = []leaderboardTrackSegment{}
	}
	clean := make([]leaderboardTrackSegment, 0, len(track.Segments))
	for _, seg := range track.Segments {
		if seg.Steps <= 0 {
			continue
		}
		if seg.Gold < 0 {
			seg.Gold = 0
		}
		if seg.Xp < 0 {
			seg.Xp = 0
		}
		clean = append(clean, seg)
	}
	track.Segments = clean
	if track.LifetimeSteps == 0 && len(track.Segments) > 0 {
		for _, seg := range track.Segments {
			track.LifetimeSteps += seg.Steps
		}
	}
	return trimLeaderboardWindow(track)
}

func appendLeaderboardTrackSegment(track leaderboardTrack, steps int, gold, xp float64) leaderboardTrack {
	if steps <= 0 {
		return track
	}
	if gold < 0 {
		gold = 0
	}
	if xp < 0 {
		xp = 0
	}
	track.Segments = append(track.Segments, leaderboardTrackSegment{
		Steps: steps,
		Gold:  gold,
		Xp:    xp,
	})
	track.LifetimeSteps += steps
	return trimLeaderboardWindow(track)
}

func trimLeaderboardWindow(track leaderboardTrack) leaderboardTrack {
	windowSteps := 0
	for _, seg := range track.Segments {
		windowSteps += seg.Steps
	}
	for windowSteps > LeaderboardWindowSteps && len(track.Segments) > 0 {
		first := track.Segments[0]
		excess := windowSteps - LeaderboardWindowSteps
		if first.Steps <= excess {
			windowSteps -= first.Steps
			track.Segments = track.Segments[1:]
			continue
		}
		keep := first.Steps - excess
		ratio := float64(keep) / float64(first.Steps)
		track.Segments[0] = leaderboardTrackSegment{
			Steps: keep,
			Gold:  first.Gold * ratio,
			Xp:    first.Xp * ratio,
		}
		windowSteps = LeaderboardWindowSteps
	}
	return track
}

func windowTotals(track leaderboardTrack) (steps int, gold, xp float64) {
	for _, seg := range track.Segments {
		steps += seg.Steps
		gold += seg.Gold
		xp += seg.Xp
	}
	return steps, gold, xp
}

func migrateLeaderboardTrackFromPlayerStats(track leaderboardTrack, stats playerStatsPayload) leaderboardTrack {
	if track.LifetimeSteps > 0 {
		return track
	}
	steps := explorationSteps(stats.CombatActionSteps, stats.RestSteps)
	if steps <= 0 {
		return track
	}
	return appendLeaderboardTrackSegment(emptyLeaderboardTrack(), steps, stats.CumulativeGold, stats.CumulativeXp)
}

func parseLeaderboardTrackFromSave(saveJSON json.RawMessage) (leaderboardTrack, playerStatsPayload, error) {
	var payload struct {
		TeamName         string          `json:"teamName"`
		PlayerStats      json.RawMessage `json:"playerStats"`
		LeaderboardTrack json.RawMessage `json:"leaderboardTrack"`
	}
	if err := json.Unmarshal(saveJSON, &payload); err != nil {
		return emptyLeaderboardTrack(), playerStatsPayload{}, err
	}
	var stats playerStatsPayload
	if len(payload.PlayerStats) > 0 {
		if err := json.Unmarshal(payload.PlayerStats, &stats); err != nil {
			return emptyLeaderboardTrack(), playerStatsPayload{}, err
		}
	}
	track := normalizeLeaderboardTrack(payload.LeaderboardTrack)
	track = migrateLeaderboardTrackFromPlayerStats(track, stats)
	return track, stats, nil
}
